manual/manual/tutorials/polymorphism.etex

```diff
@@ -0,0 +1,443 @@
+at this point, this can result in confusing type errors when later, correct
+uses are flagged as errors.
+
+\subsection{The value restriction}\label{ss:valuerestriction}
+
+Identifying the exact context in which polymorphic types should be
+replaced by weak types in a modular way is a difficult question. A first
+natural idea would be to state that any mutable object like reference, array
+or a record mutable field should never have a polymorphic type. Unfortunately,
+this rule is not sufficient, as illustrated by the following function that
+uses an internal reference to implement a delayed identity function
+\begin{caml_example}
+let make_fake_id () =
+  let store = ref None in
+  fun x -> swap store x ;;
+let fake_id = make_fake_id();;
+\end{caml_example}
+It would be unsound to apply this "fake_id" function to values with different
+types. The function "fake_id" is therefore rightfully assigned the type
+"'_a -> '_a" rather than "'a -> 'a".
+
+To circumvent this difficulty, the type checker consider that all values
+returned by a function might rely on persistent mutable states behind the
+scene and should be given a weak type. This restriction on the type of mutable
+values and function application is called the value restriction. Note that
+this value restriction is conservative: there are situations where the value
+restriction is too cautious and gives a weak type to a value that could be
+safely generalized to a polymorphic type:
+
+\begin{caml_example}
+let not_id = (fun x -> x) (fun x -> x);;
+\end{caml_example}
+
+Quite often, this happens when defining function using higher order function.
+To avoid this problem, a solution is to add an explicit argument to the
+function:
+\begin{caml_example}
+let id_again = fun x -> (fun x -> x) (fun x -> x) x;;
+\end{caml_example}
+With this argument, "id_again" is seen as a function definition by the type
+checker and can therefore be generalized. This kind of manipulation is called
+eta-expansion in lambda calculus and is sometimes referred under this name.
+
+\subsection{The relaxed value restriction}
+
+There is another partial solution to the problem of unnecessary weak type
+in the type of value, which is implemented directly within the type checker.
+Briefly, it is possible to prove that weak types that only appear as type
+parameters in covariant positions --also called positive positions-- can be
+safely generalized to polymorphic types. For instance, the type "'a  list"
+is covariant in "'a":
+\begin{caml_example}
+  let f () = [];;
+  let empty = f ();;
+\end{caml_example}
+Remark that the type inferred for "empty" is "'a list" and not "'_a list"
+that should have occurred with the value restriction since "f ()" is a
+function application.
+
+The value restriction combined with this generalization for covariant type
+parameters is called the relaxed value restriction.
+
+%question: is here the best place for describing variance?
```

Comment:
I am not sure that it is better to avoid the word variance here. I fear that any reasonable approximation would end up quite wordy:

> "As a first approximation, the generalization can also be done when the type only refers to values returned by functions that never appear as arguments of other functions or parameter of a mutable type constructor."

Is it not simpler to use the name "variance" and then explains what is variance? Note that it may be well my academic background speaking here.